core/iwasm/common/wasm_runtime_common.c

```diff
@@ -7170,3 +7178,49 @@ wasm_runtime_detect_native_stack_overflow_size(WASMExecEnv *exec_env,
     }
     return true;
 }
+
+WASM_RUNTIME_API_EXTERN bool
+wasm_runtime_is_underlying_binary_freeable(const wasm_module_inst_t module_inst)
+{
+    uint32 i;
+    bool is_freeable = false;
+
+#if WASM_ENABLE_INTERP != 0
+    if (module_inst->module_type == Wasm_Module_Bytecode) {
+#if (WASM_ENABLE_JIT != 0 || WASM_ENABLE_FAST_JIT != 0) \
+    && (WASM_ENABLE_LAZY_JIT != 0)
+        is_freeable = false;
+#elif WASM_ENABLE_FAST_INTERP == 0
+        is_freeable = false;
+#else
+        /* Fast interpreter mode */
+        WASMModule *module = ((WASMModuleInstance *)module_inst)->module;
+        is_freeable = module->is_binary_freeable;
+#if WASM_ENABLE_GC != 0 && WASM_ENABLE_STRINGREF != 0
+        is_freeable = is_freeable && module->string_literal_ptrs == NULL;
+#endif
+#if WASM_ENABLE_BULK_MEMORY != 0
+        for (i = 0; i < module->data_seg_count; i++)
+            is_freeable =
+                is_freeable
+                && bh_bitmap_get_bit(
+                    ((WASMModuleInstance *)module_inst)->e->common.data_dropped,
+                    i);
+#endif
+#endif
+    }
+#endif /* WASM_ENABLE_INTERP != 0 */
+#if WASM_ENABLE_AOT != 0
+    if (module_inst->module_type == Wasm_Module_AoT) {
+        AOTModule *module =
+            (AOTModule *)((AOTModuleInstance *)module_inst)->module;
+        is_freeable = module->is_binary_freeable;
+#if WASM_ENABLE_GC != 0 && WASM_ENABLE_STRINGREF != 0
+        is_freeable = is_freeable && module->string_literal_ptrs == NULL;
+#endif
+    }
+#endif /* WASM_ENABLE_AOT != 0 */
+
+    (void)i;
+    return is_freeable;
+}
```

Comment:
The code is a little complex, how about:
```C
WASM_RUNTIME_API_EXTERN bool
wasm_runtime_is_underlying_binary_freeable(const wasm_module_inst_t module_inst)
{
    uint32 i;

#if WASM_ENABLE_INTERP != 0
    if (module_inst->module_type == Wasm_Module_Bytecode) {
#if (WASM_ENABLE_JIT != 0 || WASM_ENABLE_FAST_JIT != 0) \
    && (WASM_ENABLE_LAZY_JIT != 0)
        return false;
#elif WASM_ENABLE_FAST_INTERP == 0
        return false;
#else
        /* Fast interpreter mode */
        WASMModule *module = ((WASMModuleInstance *)module_inst)->module;
        if (!module->is_binary_freeable)
            return false;
#if WASM_ENABLE_GC != 0 && WASM_ENABLE_STRINGREF != 0
        if (module->string_literal_ptrs)
            return false;
#endif
#if WASM_ENABLE_BULK_MEMORY != 0
        for (i = 0; i < module->data_seg_count; i++)
            if (!bh_bitmap_get_bit(
                    ((WASMModuleInstance *)module_inst)->e->common.data_dropped,
                    i))
                return fasle;
#endif
#endif
    }
#endif /* WASM_ENABLE_INTERP != 0 */
#if WASM_ENABLE_AOT != 0
    if (module_inst->module_type == Wasm_Module_AoT) {
        AOTModule *module =
            (AOTModule *)((AOTModuleInstance *)module_inst)->module;
        if (!module->is_binary_freeable)
            return false;
#if WASM_ENABLE_GC != 0 && WASM_ENABLE_STRINGREF != 0
        if (module->string_literal_ptrs)
            return false;
#endif
    }
#endif /* WASM_ENABLE_AOT != 0 */

    (void)i;
    return true;
}
```

Reply:
And could you also set `args.is_binary_freeable = false` in wasm_runtime_load, L1385?

Reply:
Right, that looks cleaner! Pushed the change